### PR TITLE
Integrate with TSYS/TSEP

### PR DESCRIPTION
**Pull request by canac**
Replace the Authorize.NET payment gateway with TSYS.

### CODE REVIEW

File: app/scripts/controllers/eventDetails.js [hunk 1]
```diff
@@ -196,7 +208,9 @@ angular.module('confRegistrationWebApp')
 
       //Credit cards
       if (_.isEmpty($scope.conference.paymentGatewayId) && _.some($scope.conference.registrantTypes, 'acceptCreditCards')) {
-        validationErrors.push('Please enter a credit card Account ID and Key under the "Payment Options" tab.');
+        var paymentGateway = $scope.paymentGateways[$scope.conference.paymentGatewayType];
+        var fields = paymentGateway ? _.map(paymentGateway.fields, 'title').join(' and ') : 'fields';
+        validationErrors.push('Please enter the credit card ' + fields + ' under the "Payment Options" tab.');
```
Comment (OzzieOrca) [inline app/scripts/controllers/eventDetails.js]: This isn't translated

File: app/scripts/controllers/eventDetails.js [hunk 2]
```diff
@@ -25,9 +25,21 @@ angular.module('confRegistrationWebApp')
...
     $scope.changeTab($scope.tabs[0]);
 
-    $scope.paymentGateways = [
-      {id: 'AUTHORIZE_NET', name: 'Authorize.Net'}
-    ];
+    $scope.paymentGateways = {
+      AUTHORIZE_NET: {
+        name: 'Authorize.Net',
+        fields: {
+          paymentGatewayId: { title: 'Account ID' },
```
Comment (OzzieOrca) [inline app/scripts/controllers/eventDetails.js]: These aren't translated

File: app/scripts/directives/payment.js [hunk 3]
```diff
@@ -58,26 +58,94 @@ angular.module('confRegistrationWebApp')
...
-                }
-                validationError = ccp.validateCardSecurityCode(currentPayment.creditCard.cvvNumber || '');
-                if(validationError){
-                  paymentErrors.push('Please enter a valid card security code. The ' + validationError + '.');
-                }
+                var validations = {
+                  nameOnCard: [
+                    {
+                      validate: required,
+                      errorMessage: 'You must enter the name on the card.'
```
Comment (OzzieOrca) [inline app/scripts/directives/payment.js]: translate

File: app/scripts/directives/payment.js [hunk 4]
```diff
@@ -58,26 +58,94 @@ angular.module('confRegistrationWebApp')
             currentPayment.transfer = currentPayment.scholarship;
           }
 
+          function required (value) {
+            return Boolean(value);
+          }
+
           var paymentErrors = [];
           if(angular.isUndefined(currentPayment.paymentType)) {
             paymentErrors.push('Please select a payment method.');
```
Comment (OzzieOrca) [inline app/scripts/directives/payment.js]: translate

File: app/scripts/directives/payment.js [hunk 5]
```diff
@@ -58,26 +58,94 @@ angular.module('confRegistrationWebApp')
...
+                        var expectedLength = cruPayments.card.info.expectedLengthForType(cardNumber) || [];
+                        var expectedDigits = expectedLength.join(gettextCatalog.getString(' or '));
+                        return gettextCatalog.getString('This is an invalid {{cardType}} number. It should have {{expectedDigits}} digits.', { cardType: cardType, expectedDigits: expectedDigits });
+                      }
+                    }, {
+                      validate: cruPayments.card.validate.checksum,
+                      errorMessage: gettextCatalog.getString('This card number is invalid. At least one digit was entered incorrectly.')
+                    }
+                  ],
+                  cvvNumber: [
```
Comment (OzzieOrca) [inline app/scripts/directives/payment.js]: In cru-payments `1.1.1` I added one more of these. https://www.npmjs.com/package/cru-payments#crupaymentscreditcardcvvvalidatecardtypelengthcvv-cardtype

American Express apparently requires the CVV to be 4 digits

File: bower.json [hunk 6]
```diff
@@ -16,6 +17,7 @@
     "pickadate": "3.5.4",
     "angular-ui-tree": "2.4.0",
     "ccp-js": "https://ccp.ccci.org/js/ccp-js.zip",
+    "cru-payments": "https://registry.npmjs.org/cru-payments/-/cru-payments-1.1.0.tgz",
```
Comment (OzzieOrca) [inline bower.json]: Can you update this to [`1.1.1`](https://github.com/CruPCI/cru-payments/releases/tag/1.1.1)?  I added American Express CVV validation. I probably should have mentioned it earlier but I was trying not to spam you with really small updates. I didn't know if there'd be more.

**Comment by canac** [comment]
@OzzieOrca I believe I have addressed all of your concerns. Does it look good now?

**Comment by OzzieOrca** [comment]
:+1

File: app/scripts/services/payment.js [hunk 7]
```diff
@@ -1,16 +1,56 @@
...
+
     // Modify a credit card payment to use a tokenized credit card instead of real credit card data
-    function tokenizeCreditCardPayment (payment) {
+    // Tokenize the credit card via Authorize.NET
+    function tokenizeCreditCardPaymentAuthorizeNet (payment) {
       return $http.get('payments/ccp-client-encryption-key').then(function (res) {
         var ccpClientEncryptionKey = res.data;
         ccp.initialize(ccpClientEncryptionKey);
         payment.creditCard.lastFourDigits = ccp.getAbbreviatedNumber(payment.creditCard.number);
         payment.creditCard.number = ccp.encrypt(payment.creditCard.number);
```
Comment (OzzieOrca) [inline app/scripts/services/payment.js]: @canac You're still encrypting the token? Is that necessary? I'm finally getting to implement this with the give site. I wasn't sure if the cvv still needed to be encrypted either. If it does, that might be nice to move into the cru-payments library.

@hlbraddock Why does the CVV need to be passed to TSYS during tokenization if the backend still needs it for a realtime transaction? If it is never needed during tokenization, that might make https://github.com/CruPCI/cru-payments/pull/16 cleaner. Is it really needed during both tokenization and when interacting with TSYS on the backend?

**Pull request merged**